pkg/nsx/services/subnet/subnet.go

```diff
@@ -0,0 +1,275 @@
+			var err error
+			if subnetService, err = InitializeSubnetService(service); err != nil {
+				log.Error(err, "failed to initialize subnet commonService")
+				os.Exit(1)
+			}
+		}
+	}
+	return subnetService
+}
+
+// InitializeSubnetService initialize Subnet service.
+func InitializeSubnetService(service common.Service) (*SubnetService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+	subnetService := &SubnetService{
+		Service: service,
+		SubnetStore: &SubnetStore{
+			ResourceStore: common.ResourceStore{
+				Indexer: cache.NewIndexer(keyFunc, cache.Indexers{
+					common.TagScopeSubnetCRUID: subnetIndexFunc,
+				}),
+				BindingType: model.VpcSubnetBindingType(),
+			},
+		},
+	}
+
+	wg.Add(1)
+	go subnetService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnet, subnetService.SubnetStore)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return subnetService, err
+	}
+	return subnetService, nil
+}
+
+// TODO Test update of VpcSubnet(eg. update tags)
+func (service *SubnetService) CreateOrUpdateSubnet(obj client.Object, orgID, projectID, vpcID string, tags []model.Tag) error {
+	uid := string(obj.GetUID())
+	nsxSubnet, err := service.buildSubnet(obj, tags)
+	if err != nil {
+		log.Error(err, "failed to build Subnet")
+		return err
+	}
+	switch obj.(type) {
+	case *v1alpha1.Subnet:
+		nsxSubnets := service.SubnetStore.GetByIndex(common.TagScopeSubnetCRUID, uid)
+		changed := false
+		if len(nsxSubnets) == 0 {
+			changed = true
+		} else {
+			changed = common.CompareResource(SubnetToComparable(&nsxSubnets[0]), SubnetToComparable(nsxSubnet))
+		}
+		if !changed {
+			log.Info("subnet not changed, skip updating", "subnet.Id", *nsxSubnets[0].Id)
+			return nil
+		}
+	}
+	orgRoot, err := service.WrapHierarchySubnet(nsxSubnet, orgID, projectID, vpcID)
+	if err != nil {
+		log.Error(err, "WrapHierarchySubnet failed")
+		return err
+	}
+	if err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam); err != nil {
+		return err
+	}
+	if err = service.CheckRealizedState(nsxSubnet, common.RealizeTimeout); err != nil {
+		return err
+	}
+	// Get Subnet from NSX after patch operation as NSX renders several fields like `path`/`parent_path`.
+	if *nsxSubnet, err = service.NSXClient.SubnetsClient.Get(orgID, projectID, vpcID, *nsxSubnet.Id); err != nil {
+		return err
+	}
+	if err = service.SubnetStore.Operate(nsxSubnet); err != nil {
+		return err
+	}
+	log.Info("successfully updated nsxSubnet", "nsxSubnet", nsxSubnet)
+	return nil
+}
+
+func (service *SubnetService) DeleteSubnet(obj client.Object, orgID, projectID, vpcID string) error {
+	nsxSubnets := service.SubnetStore.GetByIndex(common.TagScopeSubnetCRUID, string(obj.GetUID()))
+	for _, nsxSubnet := range nsxSubnets {
+		if err := service.DeleteIPAllocation(orgID, projectID, vpcID, *nsxSubnet.Id); err != nil {
```

Comment:
I do not think we should forcely delete all allocations then deleting the Subnet. If we found it returns error " it has children or it is being referenced by other objects" when deleting the VPC Subnet,  it is expected.

Reply:
Done, thanks.